File: Puissance4Game/GameLogic/GameBoard.cs
using System;
using System.Collections.Generic;

namespace Puissance4Game.GameLogic
{
    public class GameBoard
    {
        public const int Rows = 6;
        public const int Columns = 7;
        public const int PlayerOne = 1;
        public const int PlayerTwo = 2;

        private readonly int[,] _cells = new int[Rows, Columns];

        public void Reset()
        {
            Array.Clear(_cells, 0, _cells.Length);
        }

        public int this[int row, int column]
        {
            get { return _cells[row, column]; }
        }

        public bool CanDrop(int column)
        {
            return column >= 0 && column < Columns && _cells[0, column] == 0;
        }

        public IEnumerable<int> GetValidMoves()
        {
            for (var col = 0; col < Columns; col++)
            {
                if (CanDrop(col))
                {
                    yield return col;
                }
            }
        }

        public int DropPiece(int column, int player)
        {
            if (!CanDrop(column))
            {
                return -1;
            }

            for (var row = Rows - 1; row >= 0; row--)
            {
                if (_cells[row, column] == 0)
                {
                    _cells[row, column] = player;
                    return row;
                }
            }

            return -1;
        }

        public void RemovePiece(int column, int row)
        {
            if (column < 0 || column >= Columns || row < 0 || row >= Rows)
            {
                return;
            }

            if (_cells[row, column] != 0)
            {
                _cells[row, column] = 0;
            }
        }

        public bool IsFull()
        {
            for (var column = 0; column < Columns; column++)
            {
                if (_cells[0, column] == 0)
                {
                    return false;
                }
            }

            return true;
        }

        public bool HasConnectedFour(int player)
        {
            for (var row = 0; row < Rows; row++)
            {
                for (var column = 0; column < Columns; column++)
                {
                    if (_cells[row, column] == player && HasConnectedFourFrom(row, column, player))
                    {
                        return true;
                    }
                }
            }

            return false;
        }

        private bool HasConnectedFourFrom(int row, int column, int player)
        {
            foreach (var direction in Directions)
            {
                var line = CollectAlignedPositions(row, column, direction.DeltaRow, direction.DeltaColumn, player);
                if (line.Count >= 4)
                {
                    return true;
                }
            }

            return false;
        }

        public IReadOnlyList<(int Row, int Column)> GetWinningSequence(int row, int column, int player)
        {
            if (player == 0)
            {
                return Array.Empty<(int, int)>();
            }

            foreach (var direction in Directions)
            {
                var line = CollectAlignedPositions(row, column, direction.DeltaRow, direction.DeltaColumn, player);
                if (line.Count >= 4)
                {
                    return line;
                }
            }

            return Array.Empty<(int, int)>();
        }

        private static readonly (int DeltaRow, int DeltaColumn)[] Directions =
        {
            (0, 1),
            (1, 0),
            (1, 1),
            (1, -1)
        };

        private List<(int Row, int Column)> CollectAlignedPositions(int row, int column, int dr, int dc, int player)
        {
            var startRow = row;
            var startColumn = column;

            while (IsInside(startRow - dr, startColumn - dc) && _cells[startRow - dr, startColumn - dc] == player)
            {
                startRow -= dr;
                startColumn -= dc;
            }

            var positions = new List<(int Row, int Column)>();
            var currentRow = startRow;
            var currentColumn = startColumn;

            while (IsInside(currentRow, currentColumn) && _cells[currentRow, currentColumn] == player)
            {
                positions.Add((currentRow, currentColumn));
                currentRow += dr;
                currentColumn += dc;
            }

            return positions;
        }

        private static bool IsInside(int row, int column)
        {
            return row >= 0 && row < Rows && column >= 0 && column < Columns;
        }
    }
}

File: Puissance4Android/app/src/main/java/com/example/puissance4/GameBoard.java
package com.example.puissance4;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class GameBoard {
    public static final int ROWS = 6;
    public static final int COLUMNS = 7;
    public static final int PLAYER_ONE = 1;
    public static final int PLAYER_TWO = 2;

    private final int[][] cells = new int[ROWS][COLUMNS];

    public void reset() {
        for (int row = 0; row < ROWS; row++) {
            for (int column = 0; column < COLUMNS; column++) {
                cells[row][column] = 0;
            }
        }
    }

    public int getCell(int row, int column) {
        return cells[row][column];
    }

    public boolean canDrop(int column) {
        return column >= 0 && column < COLUMNS && cells[0][column] == 0;
    }

    public List<Integer> getValidMoves() {
        List<Integer> moves = new ArrayList<>();
        for (int column = 0; column < COLUMNS; column++) {
            if (canDrop(column)) {
                moves.add(column);
            }
        }
        return moves;
    }

    public int dropPiece(int column, int player) {
        if (!canDrop(column)) {
            return -1;
        }

        for (int row = ROWS - 1; row >= 0; row--) {
            if (cells[row][column] == 0) {
                cells[row][column] = player;
                return row;
            }
        }

        return -1;
    }

    public void removePiece(int column, int row) {
        if (column < 0 || column >= COLUMNS || row < 0 || row >= ROWS) {
            return;
        }

        if (cells[row][column] != 0) {
            cells[row][column] = 0;
        }
    }

    public boolean isFull() {
        for (int column = 0; column < COLUMNS; column++) {
            if (cells[0][column] == 0) {
                return false;
            }
        }
        return true;
    }

    public boolean hasConnectedFour(int player) {
        for (int row = 0; row < ROWS; row++) {
            for (int column = 0; column < COLUMNS; column++) {
                if (cells[row][column] == player && hasConnectedFourFrom(row, column, player)) {
                    return true;
                }
            }
        }
        return false;
    }

    public List<Position> getWinningSequence(int row, int column, int player) {
        if (player == 0) {
            return Collections.emptyList();
        }

        for (int[] direction : DIRECTIONS) {
            List<Position> positions = collectAlignedPositions(row, column, direction[0], direction[1], player);
            if (positions.size() >= 4) {
                return positions;
            }
        }

        return Collections.emptyList();
    }

    private boolean hasConnectedFourFrom(int row, int column, int player) {
        for (int[] direction : DIRECTIONS) {
            List<Position> positions = collectAlignedPositions(row, column, direction[0], direction[1], player);
            if (positions.size() >= 4) {
                return true;
            }
        }
        return false;
    }

    private List<Position> collectAlignedPositions(int row, int column, int dr, int dc, int player) {
        int startRow = row;
        int startColumn = column;

        while (isInside(startRow - dr, startColumn - dc) && cells[startRow - dr][startColumn - dc] == player) {
            startRow -= dr;
            startColumn -= dc;
        }

        List<Position> positions = new ArrayList<>();
        int currentRow = startRow;
        int currentColumn = startColumn;

        while (isInside(currentRow, currentColumn) && cells[currentRow][currentColumn] == player) {
            positions.add(new Position(currentRow, currentColumn));
            currentRow += dr;
            currentColumn += dc;
        }

        return positions;
    }

    private boolean isInside(int row, int column) {
        return row >= 0 && row < ROWS && column >= 0 && column < COLUMNS;
    }

    public static class Position {
        public final int row;
        public final int column;

        public Position(int row, int column) {
            this.row = row;
            this.column = column;
        }
    }

    private static final int[][] DIRECTIONS = new int[][] {
        {0, 1},
        {1, 0},
        {1, 1},
        {1, -1}
    };
}
